modules/porous_flow/src/fvkernels/FVPorousFlowDispersiveFlux.C

```diff
@@ -0,0 +1,175 @@
+    _mass_fractions(
+        getADMaterialProperty<std::vector<std::vector<Real>>>("PorousFlow_mass_frac_qp")),
+    _mass_fractions_neighbor(
+        getNeighborADMaterialProperty<std::vector<std::vector<Real>>>("PorousFlow_mass_frac_qp")),
+    _grad_mass_frac(getADMaterialProperty<std::vector<std::vector<RealGradient>>>(
+        "PorousFlow_grad_mass_frac_qp")),
+    _permeability(getADMaterialProperty<RealTensorValue>("PorousFlow_permeability_qp")),
+    _permeability_neighbor(
+        getNeighborADMaterialProperty<RealTensorValue>("PorousFlow_permeability_qp")),
+    _pressure(getADMaterialProperty<std::vector<Real>>("PorousFlow_porepressure_qp")),
+    _pressure_neighbor(
+        getNeighborADMaterialProperty<std::vector<Real>>("PorousFlow_porepressure_qp")),
+    _grad_p(getADMaterialProperty<std::vector<RealGradient>>("PorousFlow_grad_porepressure_qp")),
+    _porosity(getADMaterialProperty<Real>("PorousFlow_porosity_qp")),
+    _porosity_neighbor(getNeighborADMaterialProperty<Real>("PorousFlow_porosity_qp")),
+    _tortuosity(getADMaterialProperty<std::vector<Real>>("PorousFlow_tortuosity_qp")),
+    _tortuosity_neighbor(
+        getNeighborADMaterialProperty<std::vector<Real>>("PorousFlow_tortuosity_qp")),
+    _diffusion_coeff(
+        getMaterialProperty<std::vector<std::vector<Real>>>("PorousFlow_diffusion_coeff_qp")),
+    _diffusion_coeff_neighbor(getNeighborMaterialProperty<std::vector<std::vector<Real>>>(
+        "PorousFlow_diffusion_coeff_qp")),
+    _gravity(getParam<RealVectorValue>("gravity")),
+    _identity_tensor(ADRankTwoTensor::initIdentity),
+    _disp_long(getParam<std::vector<Real>>("disp_long")),
+    _disp_trans(getParam<std::vector<Real>>("disp_trans"))
+{
+  // Check that sufficient values of the dispersion coefficients have been entered
+  if (_disp_long.size() != _num_phases)
+    paramError(
+        "disp_long",
+        "The number of longitudinal dispersion coefficients is not equal to the number of phases");
+
+  if (_disp_trans.size() != _num_phases)
+    paramError("disp_trans",
+               "The number of transverse dispersion coefficients disp_trans is not equal to the "
+               "number of phases");
+}
+
+ADReal
+FVPorousFlowDispersiveFlux::computeQpResidual()
+{
+  ADReal flux = 0.0;
+  ADRankTwoTensor dispersion;
+  dispersion.zero();
+  ADReal diffusion;
+
+  for (unsigned int p = 0; p < _num_phases; ++p)
+  {
+    // Diffusive component
+    ADRealGradient gradX;
+
+    const auto state = determineState();
+    auto dudn = gradUDotNormal(state);
+
+    // If we are on a boundary face, use the reconstructed gradient computed in _grad_mass_frac
+    if (onBoundary(*_face_info))
+    {
+      gradX = -1 * _grad_mass_frac[_qp][p][_fluid_component];
+      diffusion = _porosity[_qp] * _tortuosity[_qp][p] * _density[_qp][p] *
+                  _diffusion_coeff[_qp][p][_fluid_component];
+    }
+    else
```

Comment:
Hi @lindsayad 

Can I get your thoughts on this bit please? If you remember from the first bit of FV that you helped me with, ghosting was removed so that there was an error when trying to access neighbour properties on a boundary element (face?). To get around it, you made a BC that added an flux BC and a Dirichlet BC (I think the logic is that using a FluxBC disables the evaluation of the kernel on the boundary face). 

However, I realised that I could use the gradient of things that we compute in materials on these edge elements (we are computing them anyway), so added the 
```
if (on boundary)
   use gradient from material
else
   compute gradient using neighbour element
```
logic you see here. This makes the FV input files just like the existing FE inputs (don't need an additional BC), and it avoids having to make a really complex BC that covers all of our cases. It seems to work ok, and I believe that if we do add a FluxBC like the one you made, that it wouldn't evaluate the `on boundary` bit due to how the FV kernels are disabled on boundaries when a FLuxBC is present.

Is my understanding correct here?

Reply:
Yes I think you're correct (a small clarification question below) and what you're doing seems logical to me. It would definitely be nicer to not have to add the additional object and have the 1-to-1 relationship with FE inputs.

So you have a Dirichlet BC right? That's the only way you will enter `computeQpResidual` for your `FVFluxKernel` object. As you correctly stated, we do not apply kernels on boundaries where FVFluxBCs reside. And we also do not apply them if *no* BCs are applied. E.g. the only time a `FVFluxKernel` runs on a boundary (outside of some forcing parameters) is if a Dirichlet BC is applied.

Reply:
OK, I think that is what I want. Use the reconstructed gradient when there is a Dirichlet BC, don't evaluate the kernel on the boundary when it is a natural boundary (no BC), and also don't evaluate the kernel on the boundary when there is a FluxBC. Thanks for the clarification.